src/main/java/com/limechain/trie/structure/nibble/Nibbles.java

```diff
@@ -0,0 +1,281 @@
+package com.limechain.trie.structure.nibble;
+
+import org.apache.commons.collections4.IteratorUtils;
+import org.apache.commons.lang3.ArrayUtils;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Objects;
+import java.util.RandomAccess;
+import java.util.stream.Collector;
+import java.util.stream.Stream;
+
+// TODO:
+//  Should this implement List or just Iterable... or maybe extend List<Nibble>?
+//  This current version contains a huge amount of boilerplate,
+//  but it's what I've reached as a most convenient option to use outside.
+//  Basically a List<Nibble> with additional methods.
+
+// TODO:
+//  As a List<Nibble> implementor, Nibbles is currently mutable, which I severely dislike,
+//  since this mandates the developer to think about where to copy and where passing by reference is sufficient.
+/**
+ * Convenience wrapper for any 'sequence of Nibble'-like structure
+ */
+public class Nibbles implements List<Nibble>, RandomAccess, Comparable<List<Nibble>> {
```

Comment:
This Nibbles class is the only remaining "big mess", since I've not yet come to a final decision for its design. Refer to the todos for the two main dilemmas, but in quick summary:
- do we want to `implements List<Nibble>`, `extends List<Nibble>` or none at all, perhaps restricting ourselves to a more generic interface like `implements Collection<Nibble>` or `implements Iterable<Nibble>` (initially, it was simply the latter, but a lot of boilerplate piled up outside); 
- do we want this to be a mutable or immutable structure? Immutability brings about a lot of benefits like eliminating the issue of ownership and the fear of outside mutations while handling a given sequence of nibbles in some part of the project. As a tradeoff, we lose the java convenience of obeying a generic interface like `List` or `Collection`, which immensely reduces boilerplate and makes the class really blissful to use... but those interfaces mandate mutability.

Reply:
I think we need a separate sync do discuss this. Is it possible to still implement `List<Nibble>` while just changing the functions to create the list as an unmodifiable one like so for example:
`Collections.unmodifiableList();` instead of `List.of();`

Reply:
> I think we need a separate sync do discuss this. Is it possible to still implement `List<Nibble>` while just changing the functions to create the list as an unmodifiable one like so for example: `Collections.unmodifiableList();` instead of `List.of();`

But the interface requires mutating methods like `public void add(int index, Nibble element)` for example... do you mean we go a similar way like Java's unmodifiable lists and simply throw runtime exceptions when attempting mutations? If so, that's an option worth considering.

Reply:
Another suggestion which I allow myself to think is viable in hindsight (coming from @stoyan-lime), is to simply abolish the whole idea for this convenience wrapper class and flow along with a simple `List<Nibble>`. 😄 Again, this will result in a lot of conversions outside, but if we manage to keep consistent use of `List` overall (instead of arrays or other data structures), this might turn out simpler... and yes, we'd lose the custom methods, but we could still go with static utility methods which might even remain in this class (thus turning `Nibbles` into a utility class).